### DISCUSSION TITLE
"Nulleinspeisung" über mehrere hm-xxx hinweg

### DISCUSSION THREAD

**Discussion by jab42** [hoylabs/OpenDTU-OnBattery#894] [Q&A]
Kann eine "Nulleinspeisung" über mehrere hm-xxx hinweg direkt über OOB gesteuert werden? Oder muss ich dazu etwas zusätzlich in Homeassistant basteln?

**Comment by spcqike** [comment]
Geht aktuell nicht. Bisher kann nur ein wechselrichter automatisch  vom DPL gesteuert werden. 

Manuell steuern (im Web UI oder auch per webrequests) kannst du alle verbundenen

**Comment by jab42** [comment]
Danke, dann kann ich also beim normalen ODTU bleiben.
Gibt es vielleicht schon eine Homeassistant Automation zu dem Zweck?
Oder irgendwo einen Thread dazu?

**Comment by genmad** [comment]
@manos1966 und ich entwickeln/testen gerade ein script für einen shelly (api gen2+) das mehrere Controller so steuert, dass diese den Hausbedarf optimal abdecken und nichts ins Netz einspeisen. Wenn das in etwa dein Usecase ist, ja dann ist da gerade etwas am entstehen.

siehe hier:  https://github.com/genmad/shellyAPIGen2AndAboveScripts/blob/master/src/openDTUonBattery/virtualPowerMeters_oDoB.js

**Reply by Manos1966** [reply]
Deutsch ist eine sehr präzise Sprache. Ich erlaube mir den Text oben zu umformulieren:
>  ich entwickle gerade ein script für einen shelly und Manos1966 macht jeden denkbaren Unsinn

😆

**Reply by jab42** [reply]
Schön. Weshalb für Shelly, also ein closed Source System und nicht z.b. für Tasmota? Die Rechenhardware wäre ja gleich, bzw bei Tasmota evtl sogar besser, wenn man zb einen esp32 mit PSRAM nimmt...

**Reply by genmad** [reply]
Weil ich einen shelly habe.
Es einfach war auf dem shelly zu realisieren.
Den Algorithmus kann man auch nach Tasmota portieren, wenn jemand Lust darauf hat.

**Reply by jab42** [reply]
Gibt es zumindest die erste beschreibung auch in deutsch?
* This script reads in the current power comsuption and 
* divides it up between n virtual powermeter. The powermeter 
* are then used to feed an controller of oDoB.
* The multiple controller will be cascaded so that at each 
* point in time only one controller is working and the others 
* are off or at their minimum power delivery. 
*
* with this all said, it is noteworthy, that all controllers 
* are connected to the same DC power supply system 
* ( solarpanels and batteries).
*

**Reply by genmad** [reply]
Der google übersetzer macht das schon gut 😏. 
Aber wie gesagt es ist noch im entstehen. Ausserdem ist das dann das falsche Forum dafür. Sollte dann in dem Projekt die Weiterentwicklung diskutiert werden.

**Comment by mastertape** [comment]
Leider wird das Thema hier nicht wirklich aufgegriffen - es gibt einige die halten es für unnötig oder zu unsicher - es geht dannn auch um die latenzen sowie platzbedarf beim programmieren .... es gibt einige dtu boards / esp32 versionnen die könnten das locker....

Es gibnt diverse ansätze zum thema z.b. gibts ne firma die sowas als externes sccript closedsource auf nen esp laufen lässt der dan die  dtu(´s) steuern würde (stichwort smart micro pv selbstbau)
Ich hoffe daher auch die ganze zeit dass sich beim thema noch was tut...

**Reply by Manos1966** [reply]
@mastertape Bin ich voll bei dir und das Thema XXL-Balkonkraftwerke bedarf viel Wissen
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/XXL-Balkonkraftwerk-1500-Watt-und-mehr%21-Worauf-Ihr-achten-solltet%21

> es geht dannn auch um die latenzen

Das ist genauso wichtig! 
wobei, @genmad s Version die wir gerade testen, hat 4Sek Reaktionszeit= von Tasmota-Stromwert-MQTTPublish bis Einstellung den neuen Limits an den Wechselrichtern.
So wird sogar @rainer667 Wasserkocher damit abgedeckt 🤣 😆

**Reply by rainer667** [reply]
Super Jungs! Dann brauche ich nur noch eine Idiotensichere Anleitung:-)

**Reply by rainer667** [reply]
@mastertape Diese Nulleinspeisung habe ich auch nur leider passiert da auch nichts mehr

**Reply by mastertape** [reply]
@Manos1966 In der Entwicklung von opendtuonbattery oder dem dpl zugrunde liegenden phyton scripts (müsste ich erst wieder mühsam suchen)  kann man nach lesen das man die tasmotas (ich glaube auch die shellys) direkt über die rest/web api abgegriffen werden. Hinergrund ist die latenz vermeidung

**Reply by Manos1966** [reply]
Meine Test-Installation:
Eine Tasmota
Eine OpenDTUonBattery#0 die per HTTP die Tasmota Daten aufnimmt und per MQTT weiterleitet.
Eine Shelly, die die MQTT Daten sammelt und per @genmad s Skript die Daten an die weitere OpenDTUonBattery berechnet.
Zwei OpenDTUonBattery, die per HTTP die Daten bei der Shelly abfragen.

Wie bereits gesagt, 4 Sekunden, von dem Moment wo die OpenDTUonBattery#0 die Tasmota Daten empfangen hat, bis das Hoymiles "Current Limit" bei sich aktualisiert hat).

Ich befürchte, die Mehrheit der Menschen die Tasmotas haben, weiss nicht Mal,  dass die schnellste Übertragungstaktung  der Tasmota 10 Sekunden ist (ausser man wiss wie man die resl-time Flagge setzten kann)...

Mit einer Latenz von 4 Sek + Tasmota auf Real-Tine kann ich gut leben.

**Reply by mastertape** [reply]
du kannst wie gesagt die mqtt werte von der opendtu on battery nehmen die zieht sie sich ja über die restapi und ist schneller als mqtt..

vielleicht auch interessant für Euch:


https://github.com/Selbstbau-PV/Selbstbau-PV-Hoymiles-nulleinspeisung-mit-OpenDTU-und-Shelly3EM (hierauf basiert dieser smart micro esp von pv selbstbau das script kann nullsteuerung meherer wr´s die haben das irgendwie auf nen esp ans laufen bekommen - aber das ist dann closed source must deren esp teil kaufen)



was wiederum hierauf basiert (achtung code veraltet)
https://gitlab.com/p3605/hoymiles-tarnkappe

Dann gibt es noch dieses projekt von jemand anderen was glaube ich derzeit öfter geupdatet ist :
https://github.com/reserve85/HoymilesZeroExport

der hammer wäre wenn man vielleicht mal überlegen würde das ganze anstatt auf nem esp irgendwie auf ne rapberry zero zu hiefen - da könnte man dan ganz andere spässe machen - oder ne opensource lösung nachbauen für den solar micro... der königsweg wäre natülich irgendwann irgendwie mehrere wr´s über die dtu on battery mit nem leistungsstarken esp in einem gerät zu regeln...

**Reply by Manos1966** [reply]
@mastertape Von den drei Projekten, die du erwähnt hast, welches kann mehrere Wechselrichter steuern?
Ich sehe nur Steuerung eines einzelnes Wechselrichters erkennen.

**Reply by schlimmchen** [reply]
Ein Shelly mit Tasmota sendet, durch die Anwendung von Regeln ("Rules") sekündlich und zuverlässig Messwerte per MQTT. Hier die Regeln aus meinem Shelly 3EM mit Tasmota:

```
rule1
ON energy#total DO publish grid/energy/total %value% ENDON
ON energy#power[1] DO publish grid/power/phase1/watts %value% ENDON
ON energy#power[2] DO publish grid/power/phase2/watts %value% ENDON
ON energy#power[3] DO publish grid/power/phase3/watts %value% ENDON
```

Das ist schon deutlich besser, als auf die Statistiken des Inverter zu warten, insbesondere wenn mehrere ins Spiel kommen, selbst wenn nur einer geregelt wird.

> Leider wird das Thema hier nicht wirklich aufgegriffen

Lese nur ich deinen Post vorwurfsvoll, oder hast du eine Erwartungshaltung, dass wir das implementieren müssten, schließlich können es andere ja auch? Das ist keine triviale Aufgabe. Das kostet Zeit. Das sind die maßgeblichen Gründe, warum es noch nicht eingebaut wurde. Dazu kommt, dass jemand, der es sich zutraut und hinreichend gut kann, auch motiviert sein muss, es zu tun.

**Reply by mastertape** [reply]
Anscheind liest in der tat nur du meinen Post missverständlich oder willst da was reininterpretieren was ich nicht gesagt habe… keine Ahnung was du da für ein Problem hast…. Ich zieh mir den Schuh den du da in den Raum wirfst jedenfalls nicht an…

**Reply by genmad** [reply]
Da muss ich dir leider widersprechen @mastertape ich fand ihn auch nicht so schön und bin da ganz bei @schlimmchen.

Ich glaube jeder will hier die Sachen vor ran bringen. Aber da alles nur auf schriftlicher Basis abläuft, da kann ich nur von mir ausgehen, sind die Posts manchmal missverständlich. Die Intentionen sind immer gut nur manchmal kommt es leider nicht so rüber. 

Deine Vorschläge von oben könnte man als begeistert von den Ideen begreifen. Aber manchmal sind Vorschläge, wenn sie unglücklich dargestellt werden und unglücklich aufgefasst werden halt nur noch Schläge und das vor ist irgendwo auf dem weg verloren gegangen.

Letzendlich ist das alles hier ein Hobby und man kann selber entscheiden was man entwickeln möchte.

**Reply by mastertape** [reply]
Es waren noch nicht mals Vorschläge - ich habe niemanden angegriffen... oder angreifen wollen  - Habre lediglich eine kurze Zusammenfassung zur Thematik  und  sowie wie (vielleicht) hilfreiche links über den tellerrand hinaus um thema liefern wollen weil ich mich gefreut habe das sich wieder jemand mit programmier kennissen am Thema interssiert ist. Wer sich daran reibt tut mir leid... Werde ich aber auch nicht weiter drauf engehen.

@Manos1966 das. Projekt von reserve85 kann wohl mehrere WR verwalten zumindest geht das aus den unteren Einträgen dieses Threads zur Config hervor:
https://github.com/reserve85/HoymilesZeroExport/issues/28

und ich meine das auch beim pv-selbstbauscribt da möglihkeit wäre - ich suche aber noch ... wenn ich es wiederfinde reiche ich nach

**Reply by schlimmchen** [reply]
@genmad Danke für dein Verständnis.

**Reply by genmad** [reply]
@schlimmchen Danke für die Zeit die du hier investierst.

**Reply by mastertape** [reply]
Dem schließe ich mich auch an seine Arbeit ist enorm... und ich habe auch nichts gegenteiliges gesagt. Mein "leider wird das Thema nicht aufgegriffen" ist keine Wertung sondern bezieht sich auf die sporadisch immer wiederkehrenden Fragen dazu, wo die Diskussion immer versandet - Das ist keine Kritik nur eine in Beobachtung aus meinen Blickwinkel - Die Gründe dafür - warum es noch nicht integriert  sind mannigfaltig, komplex, auch das habe ich erwähnt.. - und meine Hoffnung das sich vielleicht in Zukunft noch was in sachen mehrere WR dynamisch geregelt was tut,möchte ich nicht aufgeben.
Nochmal mal ich wollte niemanden auf die Füße treten - von daher  finde den Vorwurf der mir gemacht wurde nicht fair

**Reply by schlimmchen** [reply]
> von daher finde den Vorwurf der mir gemacht wurde nicht fair

Also ich hab ebenso keinen Vorwurf machen wollen. Mich hat der Kommentar getriggert, und deshalb wollte ich nachfühlen, wie das denn zu verstehen sei. Dass ich empfindlich bin ist ja gleichzeitig zutreffend, aber allein bei mir liegt es halt auch nicht, dass ich da eine Erwartungshaltung gelesen habe. Danke für deine Klarstellung.

Bei mir kommen diese Woche drei Solarmodule hinzu, die über einen weiteren Victron Laderegler die Batterie speisen, und der HM-400 kommt dann an die Batterie. Damit sind es zwei Wechselrichter an der Batterie. Und da bin ich dann motiviert, zumindest zwei Wechselrichter per DPL steuern zu können. Aber mir ist der Aufwand bewusst, und es wird noch eine ganze Weile dauern, bis das fluppt. Außerdem gibt es Themen, die ich für das Projekt als dringlich bewerte (PowerMeter Refactoring), die ich plane noch vorher anzugehen. Von mir aus darf auch gerne jemand anders schneller sein :wink:

**Reply by mastertape** [reply]
Schwamm drüber und schön das wir es klären konnten
lg

**Reply by Manos1966** [reply]
@schlimmchen wirf bitte ein Auge auf @genmad s Algorithmus:
https://github.com/genmad/shellyAPIGen2AndAboveScripts/blob/master/src/openDTUonBattery/virtualPowerMeters_oDoB.js

Ich weiss nicht ob du ihn direkt in die OpenDTUonBattery einbinden kannst (verfügbarer ESP32 Speicher). 
Die notwendigen Daten sind auf jedem Fall bereits in der OpenDTUonBattery: Stromzaehler-Consumption, Wechselrichter-Leistung, Anzahl Wechselrichter.

Er funktioniert ganz einfach: Er nimmt den Verbrauch vom Strozaehler, teilt ihn, und sendet die Teile als Powemeter Eingabe an die angeschlossenen OpenDTUs und laesst die DPL die Arbeit machen (unter Beruecksichtigung der max Leistung jeden Wecheselrichters, der min Leistung (damit sie schnell reagieren koennen) und der aktuelle Leistung).  🥇 

Ich teste ihn gerade bei mir, 1xHichiTasmota, drei OpenDTUs, 2xHoymiles und 1xShellyPlus1PM.
Er funktioniert einwandfrei und hat eine 4 Sek Reaktionszeit (Hichi-Daten Empfang, MQTT Weitergabe an die Shelly, Berechnung, Einlesen der neuen Daten von der OpenDTUs, Berechnung beim jeweiligen DPL,  Empfang des neuen "Aktuelles Limit" vom jeweiligen Wechselrichter)! 

Ich hoffe es wird dieser Woche aufhören zu regnen (Korrektur: schneien 🤦‍♂️🤦‍♂️🤦‍♂️ ), damit ich ihn bei einer Installation mit drei HM1500 auf dem Test-Gelände probieren kann (er soll bis zu fünf Wechselrichter berechnen können).

PS. Vorausgesetzt, genmad hat nichts dagegen

**Reply by genmad** [reply]
Ich habe nichts dagegen, ist ja sowieso open source. Also gerne bedienen wenn es hilfreich sein kann.

**Comment by jab42** [comment]
> Super Jungs! Dann brauche ich nur noch eine Idiotensichere Anleitung:-)

Ja, DIE brauche ich auch. :-)
Vor allem sollte die Steuerung ein Offset haben, damit Einspeisungen aus anderen Quellen, sowie gewollte Übereinspeisungen möglich sind.

**Reply by genmad** [reply]
Was meinst du mit Offset? Übereinspeisung wird wenn die Sonne scheint mal von @Manos1966 getestet.

**Reply by jab42** [reply]
Ich will eben nicht auf 0 ausregeln, sondern z.B. auf -200Watt.

**Reply by genmad** [reply]
Das könntest du in dem skript selber anpassen. Ist ja frei einsehbar und veränderbar.

**Comment by mastertape** [comment]
tasmota hätte den vorteil das man die ganzen sml zähler
(inteligente zähleinrichtung) damit abgreifen könnte -  viele haben das
schon installiert... man könnte theoretisch den wert nehmen den opentu on
battery eh abgreift wenn man schon eine nulleinspeisung damit per dpl am
laufen hat. dieser wert der im dahboard gezeigt wird  ist per mqtt
abgreifbar

Am Fr., 19. Apr. 2024 um 20:25 Uhr schrieb genmad ***@***.***

**Comment by mastertape** [comment]
Weiß jhier zufällig emand wie man diese Einstellung per mqtt oder restapi setzen kann (angestrebter Netzbezug)?
![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/42982955/ac8a926d-e597-448a-8237-c317ee3d1f55)
?

finde das nirgens in der doku oder auchh nicht im mqtt explorer

**Reply by schlimmchen** [reply]
Geht nicht. Schreib das doch bitte zu #724 dazu.

**Comment by neolauren-cmd** [comment]
Hallo,

mittlerweile mache ich die Nulleinspeisung mit einem WR über DPL und für den anderen WR benutze ich node-red auf dem Raspi. Funktioniert top.

Gruß
neo
